### ISSUE TITLE
PySet_Contains is different from set.__contains__

### ISSUE THREAD

**Issue by JelleZijlstra** [capi-workgroup/problems#6]
The documentation says "Unlike the Python __contains__() method, this function does not automatically convert unhashable sets into temporary frozensets." This isn't something that has caused me problems personally, and fixing it may not be worth the compatibility hassle, but this sort of subtle behavior difference between C and Python is a bad pattern.

The same applies to `PySet_Discard`.

**Comment by cfbolz** [comment]
hah, I had no clue about this property of `PySet_Contains` and it seems it's actually wrong in PyPy (ie our `PySet_Contains` does the conversion).

**Comment by scoder** [comment]
Yeah, I remember running into this with Cython at some point and reimplementing the missing part only then.

It seems reasonable, however, to have the C-API do simple things and avoid special cases. The fact that there is a C-API function specifically for Sets already suggests that it doesn't do exactly what e.g. `PySequence_Contains()` would do. So I'm not sure if this is a "bug". The documentation might want to spell this out as a `note`, though, or even a `warning`. Here's a PR for that:

https://github.com/python/cpython/pull/104379

**Comment by gvanrossum** [comment]
This seems to refer to [this note](https://docs.python.org/3/library/stdtypes.html#frozenset.clear):
> Note, the elem argument to the `__contains__()`, [`remove()`](https://docs.python.org/3/library/stdtypes.html#frozenset.remove), and [`discard()`](https://docs.python.org/3/library/stdtypes.html#frozenset.discard) methods may be a set. To support searching for an equivalent frozenset, a temporary one is created from elem.

This honestly feels like a wart in the Python version (although I agree that discrepancies between seemingly corresponding C and Python APIs are problematic). I'm not aware of other Python objects doing similar things. Sets are weird.

**Comment by serhiy-storchaka** [comment]
The peculiar behavior was initially introduced in 19c2d77842290af9b5f470c1eea2a71d1f77c9fe. Unfortunately there is no issue number or reference to discussion. It was documented only 5 years later in #47027, after finding issues with the then implementation. The fix made this code less efficient.

I am wondering whether arguments for it (whatever they were) are still valid.

**Comment by gvanrossum** [comment]
Let's ask @rhettinger and @aleaxit. Maybe they recall or have a mailing list reference.
